File: XGBoost.py
# XGBoost

# Install xgboost following the instructions on this link: http://xgboost.readthedocs.io/en/latest/build.html#
# Make sure xgboost in the local folder this python script is in.

# Import the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
ds = pd.read_csv("Churn_Modelling.csv")
X = ds.iloc[:, 3:13].values # [Rows, Columns] : = take all
Y = ds.iloc[:, 13].values

# Encoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X_1 = LabelEncoder() # Country
X[:, 1] = labelencoder_X_1.fit_transform(X[:, 1])
labelencoder_X_2 = LabelEncoder() # Gender
X[:, 2] = labelencoder_X_2.fit_transform(X[:, 2])
onehotencoder = OneHotEncoder(categorical_features = [1])
X = onehotencoder.fit_transform(X).toarray()
X = X[:, 1:] # Remove the first dummy variable


# Splitting the dataset into a training set and test set
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)

# Fitting XGboost to the training set
from xgboost import XGBClassifier
cl = XGBClassifier()
cl.fit(X_train, Y_train)

# Predicint the test set results
y_pred = cl.predict(X_test)
y_pred = (y_pred > 0.5) # True if larger than 0.5

# Making the confustion matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(Y_test, y_pred)

# Applying K-Fold Cross Validation
from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator = cl, X = X_train, y = Y_train, cv = 10)
accuracies.mean()
accuracies.std()

File: XGBoost.R
# XGBoost

# Importing the dataset
ds = read.csv('Churn_Modelling.csv')
ds = ds[4:14]

# Encoding categorical data
ds$Geography = as.numeric(factor(ds$Geography,
                                 levels = c('France', 'Spain', 'Germany'),
                                 labels = c(1, 2, 3)))
ds$Gender = as.numeric(factor(ds$Gender,
                              levels = c('Female', 'Male'),
                              labels = c(1, 2)))

# Splitting the dataset into a training set and test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(ds$Exited, SplitRatio = 0.8)
trs = subset(ds, split == TRUE) # Training Set
tes = subset(ds, split == FALSE) # Test Set

# Fitting XGboost to the training set
# install.packages('xgboost')
library(xgboost)

cl = xgboost(data = as.matrix(trs[-11]), label = trs$Exited, nrounds = 10)

# Applying K-Fold Cross Validation
# install.packages('caret')
library(caret)
folds = createFolds(trs$Exited, k = 10)
cv = lapply(folds, function(x) {
  trsFold = trs[-x, ]
  tesFold = trs[x, ]
  cl = xgboost(data = as.matrix(trs[-11]), label = trs$Exited, nrounds = 10)
  y_pred = predict(cl, newdata = as.matrix(tesFold[-11]))
  y_pred = (y_pred >= 0.5)
  cm = table(tesFold[, 11], y_pred)
  accuracy = (cm[1,1] + cm[2,2]) / (cm[1,1] + cm[2,2] + cm[1,2] + cm[2,1])
  return(accuracy)
})
accuracy = mean(as.numeric(cv))
